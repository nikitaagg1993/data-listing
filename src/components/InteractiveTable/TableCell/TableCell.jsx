import React from 'react';
import PropTypes from 'prop-types';
import TableCellStyle from './styled';
import { getProperty } from '../../../util';

const TableCell = ({
  id,
  column,
  row,
}) => (
  <TableCellStyle
    id={id}
    column={column}
  >
    {getProperty(row, column.id)}
  </TableCellStyle>
);

TableCell.propTypes = {
  id: PropTypes.string.isRequired,
  column: PropTypes.shape({
    id: PropTypes.string,
  }).isRequired,
  row: PropTypes.shape({}).isRequired,
};

export default TableCell;
