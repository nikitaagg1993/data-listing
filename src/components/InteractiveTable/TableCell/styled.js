import styled, { css } from 'styled-components';
// import { Cell } from './Cell';
import { media } from '../../shared/media';

const overflowCSS = css`
  div:first-child {
    white-space: ${(props) => (props.column.wrap ? 'normal' : 'nowrap')};
    overflow: ${(props) => (props.column.allowOverflow ? 'visible' : 'hidden')};
    text-overflow: ellipsis;
  }
`;

export const CellBase = styled.div`
  position: relative;
  display: flex;
  align-items: center;
  box-sizing: border-box;
  line-height: normal;
  ${({ theme, head }) => theme && theme[head ? 'headCells' : 'cells'] && theme[head ? 'headCells' : 'cells'].style};
  ${(props) => props.noPadding && 'padding: 0'};
`;

// Flex calculations
export const Cell = styled(CellBase)`
 

  ${(props) => props.column.width && css`
     width: ${props.column.width};
  `};

  ${(props) => !props.column.width && css`
    flex-grow: 1;
    flex-shrink: 0;
    flex-basis: 0;
  `};

  ${(props) => props.column.numeric && css`
    justify-content: flex-end;
  `};
  ${(props) => props.column.right && 'justify-content: flex-end'};
  ${(props) => (props.column.center || props.column.button) && 'justify-content: center'};
  ${(props) => (props.column.compact || props.column.button) && 'padding: 0'};

  /* handle hiding cells */
  ${(props) => props.column.hide && props.column.hide === 'sm' && media.sm`
    display: none;
  `};
  ${(props) => props.column.hide && props.column.hide === 'md' && media.md`
    display: none;
  `};
  ${(props) => props.column.hide && props.column.hide === 'lg' && media.lg`
    display: none;
  `};
  ${(props) => props.column.hide && Number.isInteger(props.column.hide) && media.custom(props.column.hide)`
    display: none;
  `};
`;

const TableCellStyle = styled(Cell)`
  font-size: ${(props) => props.theme.rows && props.theme.rows.fontSize};
  font-weight: 400;
  ${(props) => !props.column.cell && overflowCSS};
  ${(props) => props.column.style};
  ${(props) => props.extendedCellStyle};
`;

export default TableCellStyle;
