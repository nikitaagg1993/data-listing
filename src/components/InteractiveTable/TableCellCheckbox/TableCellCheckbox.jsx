import React, { useCallback } from 'react';
import PropTypes from 'prop-types';
import { useTableContext } from '../DataTableContext';
import Checkbox from '../Checkbox';
import TableCellCheckboxStyle from './styled';

const TableCellCheckbox = ({ name, row, selected }) => {
  const {
    dispatch,
    data,
    keyField,
  } = useTableContext();

  const handleOnRowSelected = useCallback(() => dispatch({
    type: 'SELECT_SINGLE_ROW',
    row,
    isSelected: selected,
    keyField,
    rowCount: data.length,
  }), [dispatch, row, selected, keyField, data.length]);

  return (
    <TableCellCheckboxStyle
      onClick={(e) => e.stopPropagation()}
      noPadding
    >
      <Checkbox
        name={name}
        checked={selected}
        aria-checked={selected}
        onClick={handleOnRowSelected}
      />
    </TableCellCheckboxStyle>
  );
};

TableCellCheckbox.propTypes = {
  name: PropTypes.string.isRequired,
  row: PropTypes.shape({}).isRequired,
  selected: PropTypes.bool.isRequired,
};

export default TableCellCheckbox;
