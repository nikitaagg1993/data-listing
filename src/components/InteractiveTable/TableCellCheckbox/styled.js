import styled from 'styled-components';
import { CellBase } from '../TableCell/styled';

const TableCellCheckboxStyle = styled(CellBase)`
  flex: 0 0 48px;
  justify-content: center;
  align-items: center;
  user-select: none;
  white-space: nowrap;
`;

export default TableCellCheckboxStyle;
