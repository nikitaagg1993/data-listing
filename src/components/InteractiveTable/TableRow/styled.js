import styled from 'styled-components';

const TableRowStyle = styled.div`
  display: flex;
  align-items: stretch;
  align-content: stretch;
  width: 100%;
  box-sizing: border-box;
  ${(props) => props.theme.rows && props.theme.rows.style};
  ${(props) => props.theme.rows && props.selected && props.theme.rows.selectedHighlightStyle};
  ${(props) => props.extendedRowStyle};
`;

export default TableRowStyle;
