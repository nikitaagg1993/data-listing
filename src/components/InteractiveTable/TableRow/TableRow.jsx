import React from 'react';
import PropTypes from 'prop-types';
import TableRowStyle from './styled';
import TableCell from '../TableCell';
import TableCellCheckbox from '../TableCellCheckbox';

const TableRow = ({
  id,
  keyField,
  columns,
  onRowClick,
  row,
  selectableRows,
  selected,
  selectableRowsHighlight,
  rowIndex,
}) => {
  const hightlightSelected = selectableRowsHighlight && selected;
  return (
    <>
      <TableRowStyle
        id={`row-${id}`}
        selected={hightlightSelected}
        onClick={() => onRowClick(row)}
      >
        {selectableRows && (
          <TableCellCheckbox
            name={`select-row-${row[keyField]}`}
            row={row}
            selected={selected}
          />
        )}

        {columns.map((column) => (
          <TableCell
            id={`cell-${row[keyField]}`}
            key={`cell-${Math.random()}-${row[keyField]}`}
            column={column}
            row={row}
            rowIndex={rowIndex}
          />
        ))}
      </TableRowStyle>
    </>
  );
};

TableRow.propTypes = {
  // eslint-disable-next-line react/forbid-prop-types
  id: PropTypes.any.isRequired,
  keyField: PropTypes.string.isRequired,
  columns: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
  row: PropTypes.shape({}).isRequired,
  rowIndex: PropTypes.number.isRequired,
  selectableRows: PropTypes.bool.isRequired,
  selected: PropTypes.bool.isRequired,
  selectableRowsHighlight: PropTypes.bool.isRequired,
  onRowClick: PropTypes.func.isRequired,
};

export default TableRow;
