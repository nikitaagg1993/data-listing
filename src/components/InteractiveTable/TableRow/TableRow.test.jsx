/* eslint-disable react/jsx-props-no-spreading */
import React from 'react';
import renderer from 'react-test-renderer';
import 'jest-styled-components';
import TableRow from './TableRow';
import TableCell from '../TableCell';
import TableCellCheckbox from '../TableCellCheckbox';
import TableRowStyle from './styled';

jest.mock('../DataTableContext', () => ({
  useTableContext: () => ({
    dispatch: jest.fn(),
    data: [],
    sortDirection: 'asc',
  }),
}));

const props = {
  id: 'ID',
  keyField: 'title',
  columns: [{ id: 'title', label: 'Title' }],
  onRowClick: () => null,
  row: { title: 'Amazing', name: 'Also Amazing' },
  selectableRows: true,
  selected: true,
  selectableRowsHighlight: true,
  rowIndex: 5,
};

test('Table Row', () => {
  const tree = renderer.create(<TableRow {...props} />);
  const testInstance = tree.root;

  expect(tree.toJSON()).toMatchSnapshot();
  expect(testInstance.findAllByType(TableCell).length).toEqual(1);
  expect(testInstance.findAllByType(TableCellCheckbox).length).toEqual(1);
  expect(testInstance.props).toMatchObject(props);
});

test('Table Row: Multiple Columns', () => {
  const updatedProps = {
    ...props,
    columns: [...props.columns, { id: 'name', label: 'Name' }],
  };

  const tree = renderer.create(<TableRow {...updatedProps} />);
  const testInstance = tree.root;

  expect(tree.toJSON()).toMatchSnapshot();
  expect(testInstance.findAllByType(TableCell).length).toEqual(2);
  expect(testInstance.findAllByType(TableCellCheckbox).length).toEqual(1);
  expect(testInstance.props).toMatchObject(updatedProps);
});

test('Table Row: Row not selectable', () => {
  const updatedProps = {
    ...props,
    selectableRows: false,
  };

  const tree = renderer.create(<TableRow {...updatedProps} />);
  const testInstance = tree.root;

  expect(tree.toJSON()).toMatchSnapshot();
  expect(testInstance.findAllByType(TableCell).length).toEqual(1);
  expect(testInstance.findAllByType(TableCellCheckbox).length).toEqual(0);
  expect(testInstance.props).toMatchObject(updatedProps);
});

test('Table Row: onRowClick', () => {
  const updatedProps = {
    ...props,
    onRowClick: jest.fn(() => 'row clicked'),
  };

  const tree = renderer.create(<TableRow {...updatedProps} />);
  const testInstance = tree.root;

  expect(tree.toJSON()).toMatchSnapshot();
  expect(testInstance.findAllByType(TableCell).length).toEqual(1);
  expect(testInstance.findAllByType(TableCellCheckbox).length).toEqual(1);
  expect(testInstance.findByType(TableRowStyle).props.onClick()).toBe('row clicked');
  expect(updatedProps.onRowClick).toHaveBeenCalled();
  expect(testInstance.props).toMatchObject(updatedProps);
});
