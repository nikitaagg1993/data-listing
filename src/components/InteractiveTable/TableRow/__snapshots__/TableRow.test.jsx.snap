// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table Row 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  width: 100%;
  box-sizing: border-box;
}

.c1 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-sizing: border-box;
  line-height: normal;
  padding: 0;
}

.c3 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-sizing: border-box;
  line-height: normal;
}

.c4 {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-flex-basis: 0;
  -ms-flex-preferred-size: 0;
  flex-basis: 0;
}

.c5 {
  font-weight: 400;
}

.c5 div:first-child {
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}

.c2 {
  -webkit-flex: 0 0 48px;
  -ms-flex: 0 0 48px;
  flex: 0 0 48px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  white-space: nowrap;
}

<div
  className="c0"
  id="row-ID"
  onClick={[Function]}
  selected={true}
>
  <div
    className="c1 c2"
    onClick={[Function]}
  >
    <input
      aria-label="select-row-Amazing"
      checked={true}
      name="select-row-Amazing"
      onChange={[Function]}
      onClick={[Function]}
      type="checkbox"
    />
  </div>
  <div
    className="c3 c4 c5"
    id="cell-Amazing"
  >
    Amazing
  </div>
</div>
`;

exports[`Table Row: Multiple Columns 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  width: 100%;
  box-sizing: border-box;
}

.c1 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-sizing: border-box;
  line-height: normal;
  padding: 0;
}

.c3 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-sizing: border-box;
  line-height: normal;
}

.c4 {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-flex-basis: 0;
  -ms-flex-preferred-size: 0;
  flex-basis: 0;
}

.c5 {
  font-weight: 400;
}

.c5 div:first-child {
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}

.c2 {
  -webkit-flex: 0 0 48px;
  -ms-flex: 0 0 48px;
  flex: 0 0 48px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  white-space: nowrap;
}

<div
  className="c0"
  id="row-ID"
  onClick={[Function]}
  selected={true}
>
  <div
    className="c1 c2"
    onClick={[Function]}
  >
    <input
      aria-label="select-row-Amazing"
      checked={true}
      name="select-row-Amazing"
      onChange={[Function]}
      onClick={[Function]}
      type="checkbox"
    />
  </div>
  <div
    className="c3 c4 c5"
    id="cell-Amazing"
  >
    Amazing
  </div>
  <div
    className="c3 c4 c5"
    id="cell-Amazing"
  >
    Also Amazing
  </div>
</div>
`;

exports[`Table Row: Row not selectable 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  width: 100%;
  box-sizing: border-box;
}

.c1 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-sizing: border-box;
  line-height: normal;
}

.c2 {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-flex-basis: 0;
  -ms-flex-preferred-size: 0;
  flex-basis: 0;
}

.c3 {
  font-weight: 400;
}

.c3 div:first-child {
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}

<div
  className="c0"
  id="row-ID"
  onClick={[Function]}
  selected={true}
>
  <div
    className="c1 c2 c3"
    id="cell-Amazing"
  >
    Amazing
  </div>
</div>
`;

exports[`Table Row: onRowClick 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  width: 100%;
  box-sizing: border-box;
}

.c1 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-sizing: border-box;
  line-height: normal;
  padding: 0;
}

.c3 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-sizing: border-box;
  line-height: normal;
}

.c4 {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-flex-basis: 0;
  -ms-flex-preferred-size: 0;
  flex-basis: 0;
}

.c5 {
  font-weight: 400;
}

.c5 div:first-child {
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}

.c2 {
  -webkit-flex: 0 0 48px;
  -ms-flex: 0 0 48px;
  flex: 0 0 48px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  white-space: nowrap;
}

<div
  className="c0"
  id="row-ID"
  onClick={[Function]}
  selected={true}
>
  <div
    className="c1 c2"
    onClick={[Function]}
  >
    <input
      aria-label="select-row-Amazing"
      checked={true}
      name="select-row-Amazing"
      onChange={[Function]}
      onClick={[Function]}
      type="checkbox"
    />
  </div>
  <div
    className="c3 c4 c5"
    id="cell-Amazing"
  >
    Amazing
  </div>
</div>
`;
